### PR TITLE
Sh 1932 common encryption

### PR DESCRIPTION
**Pull request by sarge**
Added support for detecting the key system for dash content within the standalone player.

### CODE REVIEW

**Comment by gscragg** [comment]
Im going to need some explanation of why you moved shit from players.service to browser.

File: app/html5.html [hunk 1]
```diff
@@ -60,26 +61,40 @@
...
-          url: qs.stream,
-          licenseServerUrl: [$location.protocol(), '://', $location.host(), '/services/license/widevine/cenc?context=', $window.btoa(qs.key)].join('')
-        }]
-      };
-
-      if (qs.type === 'dash_live'){
-        app.content.keySystem = 'com.microsoft.playready';
-        app.content.licenseServerUrl.replace('widevine', 'playready');
-      }
+      browser.mediaKeySupported('com.widevine.alpha')
```
Comment (gscragg) [inline app/html5.html]: This whole block is redundant.
Just assume widevine support and change it based on playready.

File: app/scripts/players/players.service.js [hunk 2]
```diff
@@ -114,10 +114,11 @@
     function checkDashSupport(support){
       if (support.supported){ return support; }
 
-      //chromium says it has support for widevine, but no implementations work properly
-      if (browser.isChromium()){ return support; }
+      return browser.mediaKeySupported('com.widevine.alpha').then(function(hasKey){
+
+        //chromium says it has support for widevine, but no implementations work properly
+        if (browser.isChromium()){ return support; }
```
Comment (gscragg) [inline app/scripts/players/players.service.js]: Why bother doing the media key check then checking Chromium?
Better to exit sooner.

File: app/html5.html (hunk 1 above)
Comment (sarge) [inline app/html5.html]: So the reason for moving the key checks from player to browser, is two fold.

I first investigated pulling in the player service, but that would bring all of the flash parts as well.

I moved the key system function to the browser as that is further up the dependency tree, I think detecting the key system is a good fit for the browser detection.

File: app/scripts/players/players.service.js [hunk 3]
```diff
@@ -114,10 +114,11 @@
     function checkDashSupport(support){
```
Comment (gscragg) [inline app/scripts/players/players.service.js]: I guess now there is no such thing as dash_live content, these methods `checkDashSupport` and `checkDashLiveSupport` should be renamed to match the specific keySystems.

File: app/html5.html (hunk 1 above)
Comment (sarge) [inline app/html5.html]: It think it is better to check the support of the browser rather than assume.

Comment (gscragg) [inline app/html5.html]: The browser support is already checked in admin.

File: app/scripts/players/players.service.js (hunk 2 above)
Comment (sarge) [inline app/scripts/players/players.service.js]: Sure, that's your code, but I can update that:)

Comment (gscragg) [inline app/scripts/players/players.service.js]: Check the diff.
You changed it to do the check THEN check Chromium.
Whereas it was Chromium check, then do the media key check.

Comment (sarge) [inline app/scripts/players/players.service.js]: Have updated to do the extra check for the common case

File: app/scripts/players/players.service.js (hunk 3 above)
Comment (sarge) [inline app/scripts/players/players.service.js]: Updated to reflect the key systems rather than packaging types

File: app/html5.html (hunk 1 above)
Comment (sarge) [inline app/html5.html]: Yep probably does, still doesn't mean checking it here is any less valid.

File: app/scripts/browser.service.js [hunk 4]
```diff
@@ -36,7 +36,9 @@
       isIframe:         function () { return ($window.frameElement && ($window.frameElement.nodeName === 'IFRAME')); },
 
       isFlashSupported: _isFlashSupported,
-      flashVersion:     _flashVersion
+      flashVersion:     _flashVersion,
+
+      mediaKeySupported : _mediaKeySupported
```
Comment (gscragg) [inline app/scripts/browser.service.js]: Please follow the the naming format of the other methods on this service: `isMediaKeySupported` or `isMediaKeySystemSupported`.

File: app/html5.html (hunk 1 above)
Comment (gscragg) [inline app/html5.html]: I disagree.

File: app/scripts/browser.service.js (hunk 4 above)
Comment (sarge) [inline app/scripts/browser.service.js]: Naming convention gods have been appeased isMediaKeySupported

**Comment by gscragg** [comment]
I realised at 6am this morning that most of these changes are a waste of time.
The admin system can pass in the drm type and the html5.html app can just check the querystring for "playready".

Here is where admin would send it: https://github.com/indiereign/shift72-admin/blob/master/app%2Fscripts%2Fcontent%2Fplayer%2Fplayer.js#L33

Then the html5 app can just check `qs.drmType` or similar and do what it used to do (set keySystem and change license url).

**Comment by sarge** [comment]
Sure you could also do the same for the front end player, get the calling
system to make all of the decisions. Pass it in.

My criticism of that approach is that it increases the coupling and expands
the surface of the contract between the components.

In this case the drm type does not need to be passed, as it is a feature of
the browser.

You have a video file and a browser, the player should do its best to play
the content using those inputs as independently as possible.



On Wednesday, 4 May 2016, Graham Scragg <notifications@github.com> wrote:

> I realised at 6am this morning that most of these changes are a waste of
> time.
> The admin system can pass in the drm type and the html5.html app can just
> check the querystring for "playready".
>
> Here is where admin would send it:
> https://github.com/indiereign/shift72-admin/blob/master/app%2Fscripts%2Fcontent%2Fplayer%2Fplayer.js#L33
>
> Then the html5 app can just check qs.drmType or similar and do what it
> used to do (set keySystem and change license url).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/indiereign/shift72-player/pull/15#issuecomment-216635961>
>


-- 
Mark Sargent
CHIEF TECHNOLOGY OFFICER
Level 3, 14 Garden Place
Hamilton 3204
New Zealand

*P*

+64 7 838 1114

*M*

+64 21 173 2702

<http://www.shift72.com/solo/>
<http://www.shift72.com/store/>
<http://www.shift72.com/premiere/>
<http://www.shift72.com/channel/>
<http://www.shift72.com/studio/>
www.shift72.com

**Comment by gscragg** [comment]
This specific app (html5.html) and the flash app (flash.html) are only used by the admin system. There is already some assumptions made by the admin system (should it use html5 or flash) player.

Adding the supported drm type to the constructed player url is a trivial matter.

**Comment by sarge** [comment]
Yep admin does make a bunch of assumptions about whether content can be
played or not. It does not actually know if the content will actually play.

It is easy to check the drm systems available the browser already provides
an api for that.

On Wednesday, 4 May 2016, Graham Scragg <notifications@github.com> wrote:

> This specific app (html5.html) and the flash app (flash.html) are only
> used by the admin system. There is already some assumptions made by the
> admin system (should it use html5 or flash) player.
>
> Adding the supported drm type to the constructed player url is a trivial
> matter.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/indiereign/shift72-player/pull/15#issuecomment-216641504>
>


-- 
Mark Sargent
CHIEF TECHNOLOGY OFFICER
Level 3, 14 Garden Place
Hamilton 3204
New Zealand

*P*

+64 7 838 1114

*M*

+64 21 173 2702

<http://www.shift72.com/solo/>
<http://www.shift72.com/store/>
<http://www.shift72.com/premiere/>
<http://www.shift72.com/channel/>
<http://www.shift72.com/studio/>
www.shift72.com

**Comment by sarge** [comment]
It would be worth thinking what it would be like to add fairplay support as well

**Comment by gscragg** [comment]
Begrudgingly 🌰

**Pull request merged**